File: app/src/main/kotlin/tech/mmarca/openvitals/features/manualentry/activity/ActivityPlanPickerCards.kt
package tech.mmarca.openvitals.features.manualentry.activity

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowBack
import androidx.compose.material.icons.outlined.FitnessCenter
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import tech.mmarca.openvitals.domain.model.PlannedExerciseData
import tech.mmarca.openvitals.R

@Composable
internal fun ActivityPlanActivityPickerCard(
    state: ActivityEntryUiState,
    onSelectActivity: (String) -> Unit,
    onChooseSource: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val activityTypes = state.plannedWorkouts
        .mapNotNull { it.toActivityEntryType() }
        .distinctBy { it.id }
        .sortedBy { it.id }

    Card(
        modifier = modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceContainer),
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Text(
                text = stringResource(R.string.activity_entry_plan_activity_picker_title),
                style = MaterialTheme.typography.titleMedium,
            )
            if (state.isLoadingPlannedWorkouts) {
                Text(
                    text = stringResource(R.string.activity_entry_training_plans_loading),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            } else if (activityTypes.isEmpty()) {
                Text(
                    text = stringResource(R.string.activity_entry_plan_activity_picker_empty),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            } else {
                activityTypes.forEach { type ->
                    OutlinedButton(
                        onClick = { onSelectActivity(type.id) },
                        modifier = Modifier.fillMaxWidth(),
                        shape = RoundedCornerShape(8.dp),
                    ) {
                        Icon(
                            imageVector = Icons.Outlined.FitnessCenter,
                            contentDescription = null,
                            modifier = Modifier.size(18.dp),
                        )
                        Text(
                            text = stringResource(type.labelRes),
                            modifier = Modifier.padding(start = 6.dp),
                        )
                    }
                }
            }
            OutlinedButton(
                onClick = onChooseSource,
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(8.dp),
            ) {
                Icon(
                    imageVector = Icons.AutoMirrored.Outlined.ArrowBack,
                    contentDescription = null,
                    modifier = Modifier.size(18.dp),
                )
                Text(
                    text = stringResource(R.string.activity_entry_choose_another_source),
                    modifier = Modifier.padding(start = 6.dp),
                )
            }
            state.entryError?.let { error ->
                Text(
                    text = activityEntryErrorText(error, state.detailMessage),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.error,
                )
            }
        }
    }
}

@Composable
internal fun ActivityPlanPickerCard(
    state: ActivityEntryUiState,
    onSelectPlan: (String) -> Unit,
    onChooseActivity: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val selectedTypeId = state.selectedPlannedWorkoutActivityTypeId
    val plans = state.plannedWorkouts
        .filter { plan -> plan.toActivityEntryType()?.id == selectedTypeId }
        .sortedBy { it.startTime }

    Card(
        modifier = modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.surfaceContainer),
    ) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            Text(
                text = stringResource(R.string.activity_entry_plan_picker_title),
                style = MaterialTheme.typography.titleMedium,
            )
            if (plans.isEmpty()) {
                Text(
                    text = stringResource(R.string.activity_entry_plan_picker_empty),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            } else {
                plans.forEach { plan ->
                    PlannedWorkoutButton(
                        plan = plan,
                        onClick = { onSelectPlan(plan.id) },
                    )
                }
            }
            OutlinedButton(
                onClick = onChooseActivity,
                modifier = Modifier.fillMaxWidth(),
                shape = RoundedCornerShape(8.dp),
            ) {
                Icon(
                    imageVector = Icons.AutoMirrored.Outlined.ArrowBack,
                    contentDescription = null,
                    modifier = Modifier.size(18.dp),
                )
                Text(
                    text = stringResource(R.string.activity_entry_plan_choose_activity),
                    modifier = Modifier.padding(start = 6.dp),
                )
            }
        }
    }
}

@Composable
private fun PlannedWorkoutButton(
    plan: PlannedExerciseData,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val sets = plan.toRepetitionSetInputs()
    val totalReps = sets.sumOf { set -> set.repetitionsText.toIntOrNull() ?: 0 }
    val summary = when {
        sets.isEmpty() -> stringResource(R.string.planned_workout_blocks, plan.blockCount)
        sets.size == 1 -> stringResource(R.string.activity_entry_plan_one_set_summary, totalReps)
        else -> stringResource(R.string.activity_entry_plan_summary, sets.size, totalReps)
    }

    OutlinedButton(
        onClick = onClick,
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(8.dp),
        contentPadding = PaddingValues(horizontal = 16.dp, vertical = 12.dp),
    ) {
        Column(
            modifier = Modifier.fillMaxWidth(),
            horizontalAlignment = Alignment.Start,
            verticalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            Text(
                text = plan.title ?: stringResource(R.string.activity_entry_training_plan_unnamed),
                style = MaterialTheme.typography.titleSmall,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            Text(
                text = summary,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            PlanSetPreview(sets = sets)
        }
    }
}

@Composable
private fun PlanSetPreview(sets: List<ActivityRepetitionSetInput>) {
    if (sets.isEmpty()) return
    val parts = sets.flatMapIndexed { index, set ->
        buildList {
            set.repetitionsText.toIntOrNull()?.let { reps ->
                add(stringResource(R.string.activity_entry_plan_preview_reps, reps))
            }
            set.restMinutesText.toLongOrNull()?.takeIf { it > 0L && index < sets.lastIndex }?.let { rest ->
                add(stringResource(R.string.activity_entry_plan_preview_rest, rest))
            }
        }
    }
    if (parts.isEmpty()) return

    val shown = parts.take(MaxPlanPreviewParts)
    val remaining = parts.size - shown.size
    val text = buildString {
        append(shown.joinToString(" • "))
        if (remaining > 0) {
            append(" • ")
            append(stringResource(R.string.activity_entry_plan_preview_more, remaining))
        }
    }
    Text(
        text = text,
        style = MaterialTheme.typography.bodySmall,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
        maxLines = 2,
        overflow = TextOverflow.Ellipsis,
    )
}

private const val MaxPlanPreviewParts = 5

File: lib/features/manualentry/activity/activity_plan_picker_cards.dart
import 'package:flutter/material.dart';

import '../../../domain/model/activity_models.dart';
import '../../../l10n/app_localizations.dart';
import '../../../ui/components/ov_card.dart';
import 'activity_entry_notifier.dart';
import 'activity_entry_state.dart';
import '../../../domain/model/activity_entry_types.dart';
import 'activity_entry_ui_text.dart';
import 'activity_training_plan_section.dart';

/// Port of the Kotlin `ActivityPlanPickerCards.kt`: the two steps of the
/// "create from an existing plan" flow — pick an activity type that has plans,
/// then pick one of its plans.

/// Kotlin `MaxPlanPreviewParts`.
const int _maxPlanPreviewParts = 5;

/// Kotlin `ActivityPlanActivityPickerCard`.
class ActivityPlanActivityPickerCard extends StatelessWidget {
  const ActivityPlanActivityPickerCard({
    super.key,
    required this.state,
    required this.onSelectActivity,
    required this.onChooseSource,
  });

  final ActivityEntryUiState state;
  final ValueChanged<String> onSelectActivity;
  final VoidCallback onChooseSource;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final l10n = AppLocalizations.of(context);

    final seen = <String>{};
    final activityTypes = <ActivityEntryType>[];
    for (final plan in state.plannedWorkouts) {
      final type = plannedWorkoutToActivityEntryType(plan);
      if (type != null && seen.add(type.id)) activityTypes.add(type);
    }
    activityTypes.sort((a, b) => a.id.compareTo(b.id));

    return OpenVitalsCard(
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          spacing: 12,
          children: [
            Text(l10n.activityEntryPlanActivityPickerTitle,
                style: theme.textTheme.titleMedium),
            if (state.isLoadingPlannedWorkouts)
              Text(
                l10n.activityEntryTrainingPlansLoading,
                style: theme.textTheme.bodyMedium
                    ?.copyWith(color: theme.colorScheme.onSurfaceVariant),
              )
            else if (activityTypes.isEmpty)
              Text(
                l10n.activityEntryPlanActivityPickerEmpty,
                style: theme.textTheme.bodyMedium
                    ?.copyWith(color: theme.colorScheme.onSurfaceVariant),
              )
            else
              for (final type in activityTypes)
                OutlinedButton.icon(
                  onPressed: () => onSelectActivity(type.id),
                  icon: const Icon(Icons.fitness_center_outlined, size: 18),
                  label: Text(type.label),
                ),
            OutlinedButton.icon(
              onPressed: onChooseSource,
              icon: const Icon(Icons.arrow_back, size: 18),
              label: Text(l10n.activityEntryChooseAnotherSource),
            ),
            ActivityEntryErrorText(state: state),
          ],
        ),
      ),
    );
  }
}

/// Kotlin `ActivityPlanPickerCard`.
class ActivityPlanPickerCard extends StatelessWidget {
  const ActivityPlanPickerCard({
    super.key,
    required this.state,
    required this.onSelectPlan,
    required this.onChooseActivity,
  });

  final ActivityEntryUiState state;
  final ValueChanged<String> onSelectPlan;
  final VoidCallback onChooseActivity;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final l10n = AppLocalizations.of(context);
    final plans = plannedWorkoutsForActivityType(
      state.plannedWorkouts,
      state.selectedPlannedWorkoutActivityTypeId,
    );

    return OpenVitalsCard(
      child: Padding(
        padding: const EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          spacing: 12,
          children: [
            Text(l10n.activityEntryPlanPickerTitle,
                style: theme.textTheme.titleMedium),
            if (plans.isEmpty)
              Text(
                l10n.activityEntryPlanPickerEmpty,
                style: theme.textTheme.bodyMedium
                    ?.copyWith(color: theme.colorScheme.onSurfaceVariant),
              )
            else
              for (final plan in plans)
                _PlannedWorkoutButton(
                  plan: plan,
                  onPressed: () => onSelectPlan(plan.id),
                ),
            OutlinedButton.icon(
              onPressed: onChooseActivity,
              icon: const Icon(Icons.arrow_back, size: 18),
              label: Text(l10n.activityEntryPlanChooseActivity),
            ),
          ],
        ),
      ),
    );
  }
}

class _PlannedWorkoutButton extends StatelessWidget {
  const _PlannedWorkoutButton({required this.plan, required this.onPressed});

  final PlannedExerciseData plan;
  final VoidCallback onPressed;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final l10n = AppLocalizations.of(context);
    final sets = plannedWorkoutToRepetitionSetInputs(plan);

    return OutlinedButton(
      onPressed: onPressed,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        mainAxisSize: MainAxisSize.min,
        spacing: 4,
        children: [
          Text(
            plannedWorkoutDisplayName(plan, l10n),
            maxLines: 1,
            overflow: TextOverflow.ellipsis,
            style: theme.textTheme.titleSmall,
          ),
          Text(
            plannedWorkoutSummaryText(plan, l10n),
            maxLines: 1,
            overflow: TextOverflow.ellipsis,
            style: theme.textTheme.bodyMedium
                ?.copyWith(color: theme.colorScheme.onSurfaceVariant),
          ),
          _PlanSetPreview(sets: sets),
        ],
      ),
    );
  }
}

/// Kotlin `PlanSetPreview`: "12 reps • rest 60 sec • … • +3 more".
class _PlanSetPreview extends StatelessWidget {
  const _PlanSetPreview({required this.sets});

  final List<ActivityRepetitionSetInput> sets;

  @override
  Widget build(BuildContext context) {
    if (sets.isEmpty) return const SizedBox.shrink();
    final theme = Theme.of(context);
    final l10n = AppLocalizations.of(context);

    final parts = <String>[];
    for (var index = 0; index < sets.length; index++) {
      final set = sets[index];
      final reps = int.tryParse(set.repetitionsText);
      if (reps != null) parts.add(l10n.activityEntryPlanPreviewReps(reps));
      final rest = int.tryParse(set.restMinutesText);
      // The trailing rest of the last set is not shown: nothing follows it.
      if (rest != null && rest > 0 && index < sets.length - 1) {
        parts.add(l10n.activityEntryPlanPreviewRest(rest));
      }
    }
    if (parts.isEmpty) return const SizedBox.shrink();

    final shown = parts.take(_maxPlanPreviewParts).toList();
    final remaining = parts.length - shown.length;
    final text = StringBuffer(shown.join(' • '));
    if (remaining > 0) {
      text.write(' • ${l10n.activityEntryPlanPreviewMore(remaining)}');
    }

    return Text(
      text.toString(),
      maxLines: 2,
      overflow: TextOverflow.ellipsis,
      style: theme.textTheme.bodySmall
          ?.copyWith(color: theme.colorScheme.onSurfaceVariant),
    );
  }
}
